velero-plugins/imagestream/backup.go

```diff
@@ -38,55 +40,152 @@ type BackupPlugin struct {
-		if repo == dockerRepo {
-			annotations["openshift.io/dockerImageRepo"] = repo
+	// FIXME: do we need to compare registry to dockerRepo?
+	//dockerRepo := im.Status.DockerImageRepository
+	p.Log.Info(fmt.Sprintf("internal registry: %#v", internalRegistry))
+
+	localImageCopied := false
+	localImageCopiedByTag := false
+	for _, tag := range im.Status.Tags {
+		p.Log.Info(fmt.Sprintf("tag: %#v", tag.Tag))
+		// FIXME: remove this comment once the below logic is implemented in the other plugins
+		// in imagestreamtag restore plugin: restore if not local || has a tag reference
+		// in imagestream backup plugin: copy all local images image (reverse order per tag):
+		//                                1) if tag has null tag reference, copy to dest:tag
+		//                                2) if there is a tag reference, copy to dest@sha
+		// in imagestream restore plugin: copy all local images image (reverse order per tag):
+		//                                1) if tag has null tag reference, copy to dest:tag
+		//                                2) if there is a tag reference, copy to dest@sha
+		//                                restore imagestream if no local images copied via dest tag
+		specTag := findSpecTag(im.Spec.Tags, tag.Tag)
+		copyToTag := true
+		if specTag != nil && specTag.From != nil {
+			p.Log.Info(fmt.Sprintf("image tagged: %s, %s", specTag.From.Kind, specTag.From.Name))
+			// we have a tag.
+			copyToTag = false
+		}
+		// Iterate over items in reverse order so most recently tagged is copied last
+		for i := len(tag.Items) - 1; i >= 0; i-- {
+			dockerImageReference := tag.Items[i].DockerImageReference
+			localImage := strings.HasPrefix(dockerImageReference, internalRegistry)
+			//p.Log.Info(fmt.Sprintf("image is local?: %t", localImage))
+			if localImage {
+				localImageCopied = true
+				destTag := "@" + tag.Items[i].Image
+				if copyToTag {
+					localImageCopiedByTag = true
+					destTag = ":" + tag.Tag
+				}
+				// This is where the actual image copy will go once skopeo code is vendored
+				// into plugin.
+				p.Log.Info(fmt.Sprintf("skopeo copy --src-creds=<how to get oc user:token?> --src-tls-verify=false --dest-tls-verify=false docker://%s docker://%s/%s/%s%s", dockerImageReference, migrationRegistry, im.Namespace, im.Name, destTag))
+			}
 		}
 	}
-	metadata.SetAnnotations(annotations)
+	p.Log.Info(fmt.Sprintf("copied at least one local image: %t", localImageCopied))
+	p.Log.Info(fmt.Sprintf("copied at least one local image by tag: %t", localImageCopiedByTag))
+
+	im.Annotations = annotations
+
+	var out map[string]interface{}
+	objrec, _ := json.Marshal(im)
+	json.Unmarshal(objrec, &out)
+	item.SetUnstructuredContent(out)
 
 	return item, nil, nil
 }
 
+func findSpecTag(tags []imagev1API.TagReference, name string) (*imagev1API.TagReference) {
+    for _, tag := range tags {
+        if tag.Name==name {
+            return &tag
+        }
+    }
+    return nil
+}
```

Comment:
Good catch. There's no reason for this. It happened because I pasted and modified the code from elsewhere and I didn't notice the indent problem in my editor.

Reply:
And yeah, I'd forgotten about gofmt.